File: python/benchmarks/format_comparison.py
"""Format comparison benchmarks: MeasFlow vs HDF5 (h5py).

Run:
    pip install h5py
    python benchmarks/format_comparison.py
"""

from __future__ import annotations

import os
import shutil
import statistics
import tempfile
import time

import numpy as np

from measflow.types import MeasDataType
from measflow.writer import MeasWriter
from measflow.reader import MeasReader

try:
    import h5py

    HAS_H5PY = True
except ImportError:
    HAS_H5PY = False


def _bench(fn, *, warmup: int = 1, iterations: int = 5) -> dict:
    """Run *fn* and return timing statistics."""
    for _ in range(warmup):
        fn()
    times = []
    for _ in range(iterations):
        t0 = time.perf_counter()
        fn()
        times.append(time.perf_counter() - t0)
    return {
        "median_ms": statistics.median(times) * 1000,
        "min_ms": min(times) * 1000,
        "max_ms": max(times) * 1000,
    }


class BenchmarkSuite:
    def __init__(self, sample_count: int = 1_000_000, tmp_dir: str | None = None):
        self.sample_count = sample_count
        self.tmp_dir = tmp_dir or tempfile.mkdtemp(prefix="meas_bench_")
        self.data = np.random.default_rng(42).random(sample_count, dtype=np.float32) * 10000
        self._prepare_read_files()

    def _prepare_read_files(self):
        self.meas_file = os.path.join(self.tmp_dir, "read.meas")
        self._write_measflow(self.meas_file, self.data)
        if HAS_H5PY:
            self.h5_file = os.path.join(self.tmp_dir, "read.h5")
            self._write_h5(self.h5_file, self.data)

    def cleanup(self):
        shutil.rmtree(self.tmp_dir, ignore_errors=True)

    # -- MeasFlow helpers -------------------------------------------------

    def _write_measflow(self, path: str, data: np.ndarray):
        with MeasWriter(path) as w:
            g = w.add_group("Data")
            ch = g.add_channel("Signal", MeasDataType.Float32, track_statistics=False)
            ch.write_bulk(data)

    def _read_measflow(self, path: str) -> np.ndarray:
        with MeasReader(path) as r:
            return r["Data"]["Signal"].read_all()

    # -- HDF5 helpers -----------------------------------------------------

    def _write_h5(self, path: str, data: np.ndarray):
        with h5py.File(path, "w") as f:
            grp = f.create_group("Data")
            grp.create_dataset("Signal", data=data)

    def _read_h5(self, path: str) -> np.ndarray:
        with h5py.File(path, "r") as f:
            return f["Data"]["Signal"][:]

    # -- Benchmarks -------------------------------------------------------

    def bench_write_1ch(self) -> dict:
        results = {}
        path = os.path.join(self.tmp_dir, "w1.meas")
        results["MeasFlow"] = _bench(lambda: self._write_measflow(path, self.data))
        results["MeasFlow"]["size_kb"] = os.path.getsize(path) / 1024
        if HAS_H5PY:
            path_h5 = os.path.join(self.tmp_dir, "w1.h5")
            results["HDF5 (h5py)"] = _bench(lambda: self._write_h5(path_h5, self.data))
            results["HDF5 (h5py)"]["size_kb"] = os.path.getsize(path_h5) / 1024
        return results

    def bench_write_10ch(self) -> dict:
        results = {}

        def write_meas_10ch():
            p = os.path.join(self.tmp_dir, "w10.meas")
            with MeasWriter(p) as w:
                g = w.add_group("Data")
                for c in range(10):
                    ch = g.add_channel(f"Ch{c}", MeasDataType.Float32, track_statistics=False)
                    ch.write_bulk(self.data)

        results["MeasFlow"] = _bench(write_meas_10ch)

        if HAS_H5PY:
            def write_h5_10ch():
                p = os.path.join(self.tmp_dir, "w10.h5")
                with h5py.File(p, "w") as f:
                    grp = f.create_group("Data")
                    for c in range(10):
                        grp.create_dataset(f"Ch{c}", data=self.data)

            results["HDF5 (h5py)"] = _bench(write_h5_10ch)
        return results

    def bench_read_1ch(self) -> dict:
        results = {}
        results["MeasFlow"] = _bench(lambda: self._read_measflow(self.meas_file))
        if HAS_H5PY:
            results["HDF5 (h5py)"] = _bench(lambda: self._read_h5(self.h5_file))
        return results

    def bench_read_10ch(self) -> dict:
        results = {}

        # Prepare 10-channel files
        meas10 = os.path.join(self.tmp_dir, "r10.meas")
        with MeasWriter(meas10) as w:
            g = w.add_group("Data")
            for c in range(10):
                ch = g.add_channel(f"Ch{c}", MeasDataType.Float32, track_statistics=False)
                ch.write_bulk(self.data)

        def read_meas_10ch():
            with MeasReader(meas10) as r:
                for c in range(10):
                    r["Data"][f"Ch{c}"].read_all()

        results["MeasFlow"] = _bench(read_meas_10ch)

        if HAS_H5PY:
            h510 = os.path.join(self.tmp_dir, "r10.h5")
            with h5py.File(h510, "w") as f:
                grp = f.create_group("Data")
                for c in range(10):
                    grp.create_dataset(f"Ch{c}", data=self.data)

            def read_h5_10ch():
                with h5py.File(h510, "r") as f:
                    for c in range(10):
                        f["Data"][f"Ch{c}"][:]

            results["HDF5 (h5py)"] = _bench(read_h5_10ch)
        return results

    def bench_streaming_write(self) -> dict:
        results = {}
        chunk_size = self.sample_count // 10

        def stream_meas():
            p = os.path.join(self.tmp_dir, "stream.meas")
            with MeasWriter(p) as w:
                g = w.add_group("Data")
                ch = g.add_channel("Signal", MeasDataType.Float32, track_statistics=False)
                for i in range(10):
                    ch.write_bulk(self.data[i * chunk_size : (i + 1) * chunk_size])
                    w.flush()

        # HDF5 has no streaming support — MeasFlow only
        results["MeasFlow (10 flushes)"] = _bench(stream_meas)
        return results

    def bench_file_size(self) -> dict:
        results = {}
        p_meas = os.path.join(self.tmp_dir, "size.meas")
        self._write_measflow(p_meas, self.data)
        results["MeasFlow"] = {"size_kb": os.path.getsize(p_meas) / 1024}

        if HAS_H5PY:
            p_h5 = os.path.join(self.tmp_dir, "size.h5")
            self._write_h5(p_h5, self.data)
            results["HDF5 (h5py)"] = {"size_kb": os.path.getsize(p_h5) / 1024}

        p_raw = os.path.join(self.tmp_dir, "size.bin")
        with open(p_raw, "wb") as f:
            f.write(self.data.tobytes())
        results["Raw binary"] = {"size_kb": os.path.getsize(p_raw) / 1024}
        return results


def _print_results(name: str, results: dict):
    print(f"\n{'-' * 60}")
    print(f"  {name}")
    print(f"{'-' * 60}")
    for label, data in results.items():
        if "median_ms" in data:
            print(f"  {label}: {data['median_ms']:8.2f} ms")
        elif "size_kb" in data:
            print(f"  {label}: {data['size_kb']:8.1f} KB")


def main():
    for n in [100_000, 1_000_000]:
        print(f"\n{'=' * 60}")
        print(f"  Format comparison (Python) -- {n} samples")
        print(f"{'=' * 60}")

        suite = BenchmarkSuite(sample_count=n)
        try:
            _print_results("Write 1 channel", suite.bench_write_1ch())
            _print_results("Write 10 channels", suite.bench_write_10ch())
            _print_results("Read 1 channel", suite.bench_read_1ch())
            _print_results("Read 10 channels", suite.bench_read_10ch())
            _print_results("Streaming write", suite.bench_streaming_write())
            _print_results("File size", suite.bench_file_size())
        finally:
            suite.cleanup()

    if not HAS_H5PY:
        print("\nWarning: h5py not installed -- only MeasFlow results shown.")
        print("  Install with: pip install h5py")


if __name__ == "__main__":
    main()

File: c/benchmarks/format_comparison.c
/**
 * Format comparison benchmarks: MeasFlow vs HDF5 (libhdf5).
 *
 * Build (with vcpkg):
 *   vcpkg install hdf5
 *   cmake -B build -S c -DMEAS_BUILD_BENCHMARKS=ON \
 *     -DCMAKE_TOOLCHAIN_FILE=$VCPKG_INSTALLATION_ROOT/scripts/buildsystems/vcpkg.cmake
 *   cmake --build build --config Release
 *   ./build/bench_format_comparison    (or build/Release/bench_format_comparison.exe)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "../measflow.h"

#ifdef MEAS_HAVE_HDF5
#include <hdf5.h>
#endif

/* ── Timing helpers ─────────────────────────────────────────────────────── */

#ifdef _WIN32
#include <windows.h>
static double now_ms(void)
{
    LARGE_INTEGER freq, cnt;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&cnt);
    return (double)cnt.QuadPart / (double)freq.QuadPart * 1000.0;
}
#else
static double now_ms(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}
#endif

typedef struct {
    double median_ms;
    double min_ms;
    double max_ms;
} BenchResult;

static int cmp_double(const void *a, const void *b)
{
    double da = *(const double *)a, db = *(const double *)b;
    return (da > db) - (da < db);
}

static BenchResult bench(void (*fn)(const char *, const float *, int),
                          const char *path, const float *data, int n,
                          int warmup, int iterations)
{
    int i;
    for (i = 0; i < warmup; i++)
        fn(path, data, n);

    double *times = (double *)malloc((size_t)iterations * sizeof(double));
    for (i = 0; i < iterations; i++) {
        double t0 = now_ms();
        fn(path, data, n);
        times[i] = now_ms() - t0;
    }
    qsort(times, (size_t)iterations, sizeof(double), cmp_double);

    BenchResult r;
    r.median_ms = times[iterations / 2];
    r.min_ms = times[0];
    r.max_ms = times[iterations - 1];
    free(times);
    return r;
}

/* ── MeasFlow write/read ────────────────────────────────────────────────── */

static void write_measflow(const char *path, const float *data, int n)
{
    MeasWriter *w = meas_writer_open(path);
    MeasGroupWriter *g = meas_writer_add_group(w, "Data");
    MeasChannelWriter *ch = meas_group_add_channel(g, "Signal", MEAS_FLOAT32);
    meas_channel_set_statistics(ch, 0);
    meas_channel_write_f32(ch, data, n);
    meas_writer_close(w);
}

static void read_measflow(const char *path, const float *data, int n)
{
    (void)data;
    MeasReader *r = meas_reader_open(path);
    if (!r) return;
    const MeasGroupData *grp = meas_reader_group_by_name(r, "Data");
    if (grp) {
        const MeasChannelData *ch = meas_group_channel_by_name(grp, "Signal");
        if (ch) {
            float *buf = (float *)malloc((size_t)n * sizeof(float));
            meas_channel_read_f32(ch, buf, n);
            free(buf);
        }
    }
    meas_reader_close(r);
}

static void write_measflow_10ch(const char *path, const float *data, int n)
{
    MeasWriter *w = meas_writer_open(path);
    MeasGroupWriter *g = meas_writer_add_group(w, "Data");
    char name[16];
    for (int c = 0; c < 10; c++) {
        snprintf(name, sizeof(name), "Ch%d", c);
        MeasChannelWriter *ch = meas_group_add_channel(g, name, MEAS_FLOAT32);
        meas_channel_set_statistics(ch, 0);
        meas_channel_write_f32(ch, data, n);
    }
    meas_writer_close(w);
}

static void read_measflow_10ch(const char *path, const float *data, int n)
{
    (void)data;
    MeasReader *r = meas_reader_open(path);
    if (!r) return;
    const MeasGroupData *grp = meas_reader_group_by_name(r, "Data");
    if (grp) {
        char name[16];
        float *buf = (float *)malloc((size_t)n * sizeof(float));
        for (int c = 0; c < 10; c++) {
            snprintf(name, sizeof(name), "Ch%d", c);
            const MeasChannelData *ch = meas_group_channel_by_name(grp, name);
            if (ch) meas_channel_read_f32(ch, buf, n);
        }
        free(buf);
    }
    meas_reader_close(r);
}

static void stream_measflow(const char *path, const float *data, int n)
{
    MeasWriter *w = meas_writer_open(path);
    MeasGroupWriter *g = meas_writer_add_group(w, "Data");
    MeasChannelWriter *ch = meas_group_add_channel(g, "Signal", MEAS_FLOAT32);
    meas_channel_set_statistics(ch, 0);
    int chunk = n / 10;
    int i;
    for (i = 0; i < 10; i++) {
        meas_channel_write_f32(ch, data + i * chunk, chunk);
        meas_writer_flush(w);
    }
    meas_writer_close(w);
}

/* ── HDF5 write/read ────────────────────────────────────────────────────── */

#ifdef MEAS_HAVE_HDF5
static void write_hdf5(const char *path, const float *data, int n)
{
    hid_t file = H5Fcreate(path, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
    hid_t grp = H5Gcreate2(file, "Data", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    hsize_t dims[1] = { (hsize_t)n };
    hid_t space = H5Screate_simple(1, dims, NULL);
    hid_t dset = H5Dcreate2(grp, "Signal", H5T_IEEE_F32LE, space,
                              H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    H5Dwrite(dset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, data);
    H5Dclose(dset);
    H5Sclose(space);
    H5Gclose(grp);
    H5Fclose(file);
}

static void read_hdf5(const char *path, const float *data, int n)
{
    (void)data;
    hid_t file = H5Fopen(path, H5F_ACC_RDONLY, H5P_DEFAULT);
    hid_t dset = H5Dopen2(file, "/Data/Signal", H5P_DEFAULT);
    float *buf = (float *)malloc((size_t)n * sizeof(float));
    H5Dread(dset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, buf);
    free(buf);
    H5Dclose(dset);
    H5Fclose(file);
}

static void read_hdf5_10ch(const char *path, const float *data, int n)
{
    (void)data;
    hid_t file = H5Fopen(path, H5F_ACC_RDONLY, H5P_DEFAULT);
    float *buf = (float *)malloc((size_t)n * sizeof(float));
    char name[32];
    for (int c = 0; c < 10; c++) {
        snprintf(name, sizeof(name), "/Data/Ch%d", c);
        hid_t dset = H5Dopen2(file, name, H5P_DEFAULT);
        H5Dread(dset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, buf);
        H5Dclose(dset);
    }
    free(buf);
    H5Fclose(file);
}

static void write_hdf5_10ch(const char *path, const float *data, int n)
{
    hid_t file = H5Fcreate(path, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
    hid_t grp = H5Gcreate2(file, "Data", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    hsize_t dims[1] = { (hsize_t)n };
    hid_t space = H5Screate_simple(1, dims, NULL);
    char name[16];
    for (int c = 0; c < 10; c++) {
        snprintf(name, sizeof(name), "Ch%d", c);
        hid_t dset = H5Dcreate2(grp, name, H5T_IEEE_F32LE, space,
                                  H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
        H5Dwrite(dset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, data);
        H5Dclose(dset);
    }
    H5Sclose(space);
    H5Gclose(grp);
    H5Fclose(file);
}

static void stream_hdf5(const char *path, const float *data, int n)
{
    /* HDF5 does not support incremental writes without extensible datasets.
       Write all data at once for a fair "best case" comparison. */
    write_hdf5(path, data, n);
}
#endif

/* ── File size helper ───────────────────────────────────────────────────── */

static long file_size(const char *path)
{
    FILE *f = fopen(path, "rb");
    if (!f) return 0;
    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fclose(f);
    return sz;
}

/* ── Print helpers ──────────────────────────────────────────────────────── */

static void print_header(const char *title)
{
    printf("\n");
    for (int i = 0; i < 60; i++) putchar('-');
    printf("\n  %s\n", title);
    for (int i = 0; i < 60; i++) putchar('-');
    printf("\n");
}

static void print_result(const char *label, BenchResult r)
{
    printf("  %s: %8.2f ms\n", label, r.median_ms);
}

static void print_size(const char *label, long bytes)
{
    printf("  %s: %8.1f KB\n", label, bytes / 1024.0);
}

/* ── Main ───────────────────────────────────────────────────────────────── */

int main(void)
{
    int sample_counts[] = { 100000, 1000000 };
    int num_counts = 2;
    int s;

    for (s = 0; s < num_counts; s++) {
        int n = sample_counts[s];

        /* Generate test data */
        float *data = (float *)malloc((size_t)n * sizeof(float));
        srand(42);
        for (int i = 0; i < n; i++)
            data[i] = (float)rand() / (float)RAND_MAX * 10000.0f;

        char meas_path[256], h5_path[256], raw_path[256];
        snprintf(meas_path, sizeof(meas_path), "bench_%d.meas", n);
        snprintf(h5_path, sizeof(h5_path), "bench_%d.h5", n);
        snprintf(raw_path, sizeof(raw_path), "bench_%d.bin", n);

        printf("\n============================================================\n");
        printf("  Format comparison (C) -- %d samples\n", n);
        printf("============================================================\n");

        /* Write benchmarks */
        print_header("Write 1 channel");
        print_result("MeasFlow", bench(write_measflow, meas_path, data, n, 1, 5));
#ifdef MEAS_HAVE_HDF5
        print_result("HDF5 (libhdf5)", bench(write_hdf5, h5_path, data, n, 1, 5));
#endif

        /* Write 10 channels */
        print_header("Write 10 channels");
        {
            char meas10_path[256], h510_path[256];
            snprintf(meas10_path, sizeof(meas10_path), "bench_%d_10ch.meas", n);
            snprintf(h510_path, sizeof(h510_path), "bench_%d_10ch.h5", n);
            print_result("MeasFlow", bench(write_measflow_10ch, meas10_path, data, n, 1, 5));
#ifdef MEAS_HAVE_HDF5
            print_result("HDF5 (libhdf5)", bench(write_hdf5_10ch, h510_path, data, n, 1, 5));
#endif
            remove(meas10_path);
            remove(h510_path);
        }

        /* Read benchmarks */
        write_measflow(meas_path, data, n);
#ifdef MEAS_HAVE_HDF5
        write_hdf5(h5_path, data, n);
#endif

        print_header("Read 1 channel");
        print_result("MeasFlow", bench(read_measflow, meas_path, data, n, 1, 5));
#ifdef MEAS_HAVE_HDF5
        print_result("HDF5 (libhdf5)", bench(read_hdf5, h5_path, data, n, 1, 5));
#endif

        /* Read 10 channels */
        print_header("Read 10 channels");
        {
            char meas10_path[256], h510_path[256];
            snprintf(meas10_path, sizeof(meas10_path), "bench_%d_10ch.meas", n);
            snprintf(h510_path, sizeof(h510_path), "bench_%d_10ch.h5", n);
            write_measflow_10ch(meas10_path, data, n);
            print_result("MeasFlow", bench(read_measflow_10ch, meas10_path, data, n, 1, 5));
#ifdef MEAS_HAVE_HDF5
            write_hdf5_10ch(h510_path, data, n);
            print_result("HDF5 (libhdf5)", bench(read_hdf5_10ch, h510_path, data, n, 1, 5));
#endif
            remove(meas10_path);
            remove(h510_path);
        }

        /* Streaming write (MeasFlow only — HDF5 has no streaming support) */
        print_header("Streaming write");
        print_result("MeasFlow (10 flushes)", bench(stream_measflow, meas_path, data, n, 1, 5));

        /* File size */
        write_measflow(meas_path, data, n);
        print_header("File size");
        print_size("MeasFlow", file_size(meas_path));
#ifdef MEAS_HAVE_HDF5
        write_hdf5(h5_path, data, n);
        print_size("HDF5 (libhdf5)", file_size(h5_path));
#endif
        /* Raw binary baseline */
        {
            FILE *f = fopen(raw_path, "wb");
            fwrite(data, sizeof(float), (size_t)n, f);
            fclose(f);
            print_size("Raw binary", file_size(raw_path));
        }

        /* Cleanup */
        remove(meas_path);
        remove(h5_path);
        remove(raw_path);
        free(data);
    }

#ifndef MEAS_HAVE_HDF5
    printf("\n  Note: HDF5 not available. Install via: vcpkg install hdf5\n");
    printf("  Build with -DMEAS_BUILD_BENCHMARKS=ON to include HDF5 comparison.\n");
#endif

    return 0;
}
